### PR TITLE
4830 debug tiles

### LINKED ISSUE #4830: showTileBoundaries debug option should read from a source whose min/maxzoom include current zoom if available
**mapbox-gl-js version**: 0.38.0

### Steps to Trigger Behavior

 1. Visit [the openmaptiles style page](https://openmaptiles.org/styles/), open the javascript console.
 2. Paste `maps['osm-liberty'].showTileBoundaries = true;` in the console.
 3. Note the red tile boundaries appear in osm-liberty map on that page. If you zoom that map beyond zoom 6 the tile boundaries disappear.

### Expected Behavior

The boundary tiles should not disappear.

### Actual Behavior

The boundary tiles disappear.

While the instructions above reference a page that's using mapbox-gl 0.31.0, I have a private test case that's using 0.38.0 that's exhibiting the same behavior. Unfortunately I can't link to it because it's behind firewalls.

I've opened a ticket on the style maker's github here: lukasmartinelli/osm-liberty#13. Here's a relevant excerpt from that ticket:

I found this out because I've copied this style to a map I'm working on as a proof-of-concept for the use of mapbox-gl for my company. I've noticed this problem is tied to the natural_earth layer in style.json. If I completely remove this layer definition and reload the page, setting `showTileBoundaries = true` does absolutely nothing. If I remove the source declaration for `natural_earth_shaded_relief` after removing the `natural_earth` layer, everything works as expected.

### LINKED ISSUE #5535: overzooming tiles in studio not reporting correct zxy in deubg mode
**mapbox-gl-js version**: latest in mapbox-studio 

studio-zoom

in studio when you turn on debug mode the z/x/y helper doesn't update after z16;  guessing it is reflecting the data that is being displayed, and at z16+ its still displaying data from z16 mvt. which makes sense, but showing a corresponding z## with that x,y coordinate can cause confusing.

dropping ticket here, per @samanpwbb this is a bug in gl-js and not studio 

### Steps to Trigger Behavior

 1. turn on debug->tile boundaries in Mapbox studio 
 2. zoom in past z16+

### Expected Behavior
debug's tile coordinates for x,y at z16+ reflect the z16 tile

### Actual Behavior
label debug tiles with z16 consistently since that's the vt's data that's being show or to update to proper z-x-y coordinates in debug mode

### PR DESCRIPTION
**Pull request by ryanhamley**
This PR implements two improvements to debug tile data:
- Use the source with the highest `maxzoom` so that more granular tile debug data is always available
- Describe the actual and overscale tile ID on debug tiles

Closes #4830
Closes #5535

This re-opens https://github.com/mapbox/mapbox-gl-js/pull/4990. Somehow when trying to rebase that PR, I accidentally removed the commits and closed out #4990 🤷‍♂️  Please refer to that PR for discussion and code review.

### CODE REVIEW

File: src/render/draw_debug.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ function drawDebugTile(painter, sourceCache, coord) {
         debugUniformValues(posMatrix, Color.red), id,
         painter.debugBuffer, painter.tileBorderIndexBuffer, painter.debugSegments);
 
-    const vertices = createTextVertices(coord.toString(), 50, 200, 5);
+    const vertices = createTextVerticies(coord.canonical.toString(), 50, 200, 5);
```
Comment (asheemmamoowala) [inline src/render/draw_debug.js]: typo: `createTextVertices`

File: src/render/painter.js [hunk 2]
```diff
@@ -395,9 +397,22 @@ class Painter {
...
-            for (const id in sourceCaches) {
-                draw.debug(this, sourceCaches[id], coordsAscending[id]);
-                break;
+            //Use source with highest maxzoom
+            let selectedSource;
+            let sourceCache;
+            const layers = values(this.style._layers);
+            layers.forEach((layer) => {
+                if (layer.source && !layer.isHidden(this.transform.zoom)) {
+                    if (layer.source !== (sourceCache && sourceCache.id)) {
```
Comment (asheemmamoowala) [inline src/render/painter.js]: nit: the null check for `sourceCache` makes this code a bit hard to read.

**Comment by andrewharvey** [comment]
This is closely related to https://github.com/mapbox/mapbox-gl-js/issues/8376#issuecomment-504619187, which also would create conflicts. I'm happy to help rebase this so these changes can be applied cleanly but it might make sense to wait for #8380.

**Comment by asheemmamoowala** [comment]
@andrewharvey Want to take the changes from this into #8380 and help close this out ?

**Comment by andrewharvey** [comment]
Thanks for refactoring this @asheemmamoowala, it looks good to me from just reading the code.

**Comment by vakila** [comment]
Hi folks, just jumping in here to check in as part of some PR gardening we're doing.

@ryanhamley @andrewharvey Where does this stand at the moment? AFAICT it's not waiting on any requested changes, so is it in need of another round of review, or ready to merge?

File: src/render/painter.js [hunk 3]
```diff
@@ -430,9 +431,22 @@ class Painter {
...
 
         if (this.options.showTileBoundaries) {
-            for (const id in sourceCaches) {
-                draw.debug(this, sourceCaches[id], coordsAscending[id]);
-                break;
+            //Use source with highest maxzoom
+            let selectedSource;
+            let sourceCache;
+            const layers = values(this.style._layers);
+            layers.forEach((layer) => {
```
Comment (arindam1993) [inline src/render/painter.js]: nit: Use a `for..of`?

Also would it be possible to clean up the loop somehow? seems its a bit unnecessarily complex for what it needs to do

Comment (ryanhamley) [reply src/render/painter.js]: Simplified it a good bit 👍

**Comment by vakila** [comment]
@ryanhamley Thanks for the update! Looks like the `debug/tile` render test is now failing (I tried rerunning the tests in case it was a fluke but got the same failure), and there's also a small lint error to be fixed. I'll leave it to @arindam1993 to speak to the latest changes per his comment.

**Comment by ryanhamley** [comment]
Yep, working on those now. Not sure why my change would affect the test

**Comment by ryanhamley** [comment]
I tried a couple approaches to simplifying the loop and they don't really work well. I ended up having to add back all of the same checks that are in the current PR. I think it's fine as is so I'm going with the original PR.

File: src/render/painter.js (hunk 3 above)
Comment (vakila) [reply src/render/painter.js]: Looks like this has become moot as we've reverted to the original `forEach`, but for future reference just noting here that IE can't handle `for...of` :grimacing:

**Pull request merged**